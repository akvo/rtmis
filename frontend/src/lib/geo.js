import { feature, merge } from "topojson-client";
import { geoCentroid, geoBounds } from "d3-geo";

const topojson = window.topojson;
const topojson_object = topojson.objects[Object.keys(topojson.objects)[0]];
const shapeLevels = Object.keys(topojson_object.geometries[0].properties);

const tile = {
  url: "https://server.arcgisonline.com/ArcGIS/rest/services/Ocean_Basemap/MapServer/tile/{z}/{y}/{x}",
  attribution:
    "Tiles &copy; Esri &mdash; Sources: GEBCO, NOAA, CHS, OSU, UNH, CSUMB, National Geographic, DeLorme, NAVTEQ, and Esri",
};

const getBounds = (administration) => {
  const selected = administration.map((x, i) => {
    return {
      value: x.name,
      prop: shapeLevels[i],
    };
  });
  const geoFilter = topojson_object.geometries.filter((x) => {
    const filters = [];
    selected.forEach((s) => {
      if (x?.properties?.[s.prop] === s.value) {
        filters.push(true);
      } else {
        filters.push(false);
      }
    });
    return filters?.filter((f) => f).length === selected.length;
  });
  const mergeTopo = merge(
    topojson,
    geoFilter.length ? geoFilter : topojson_object.geometries
  );
  const center = geoCentroid(mergeTopo).reverse();
  const bounds = geoBounds(mergeTopo);
  const bbox = [bounds[0].reverse(), bounds[1].reverse()];
  return {
    coordinates: center,
    bbox: bbox,
  };
};

const defaultPos = () => {
  const mergeTopo = merge(topojson, topojson_object.geometries);
  const center = geoCentroid(mergeTopo).reverse();
  const bounds = geoBounds(mergeTopo);
  const bbox = [bounds[0].reverse(), bounds[1].reverse()];
  return {
    coordinates: center,
    bbox: bbox,
  };
};

const geojson = feature(topojson, topojson_object);

const geo = {
  geojson: geojson,
  shapeLevels: shapeLevels,
  tile: tile,
  getBounds: getBounds,
  defaultPos: defaultPos,
};

export default geo;
