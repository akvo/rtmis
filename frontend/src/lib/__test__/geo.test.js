import "@testing-library/jest-dom";
import geo from "../geo";

const administration = [
  {
    childLevelName: "County",
    children: [
      {
        full_name: "Kenya|Baringo",
        id: 2,
        level: 1,
        name: "Baringo",
        parent: 1,
        path: "1.",
      },
    ],
    full_name: "Kenya",
    id: 1,
    level: 0,
    levelName: "National",
    name: "Kenya",
    parent: null,
    path: null,
  },
];

describe("geo", () => {
  test("test if getBounds when empty array is passed as administration", () => {
    const expectedResult = {
      bbox: [
        [NaN, NaN],
        [NaN, NaN],
      ],
      coordinates: [NaN, NaN],
    };
    expect(geo.getBounds([])).toEqual(expectedResult);
  });
  test("test if getBounds with an administration", () => {
    const bounds = geo.getBounds(administration);
    expect(typeof bounds).toBe("object");
    expect(bounds).toHaveProperty("coordinates");
  });
  test("defaultPos", () => {
    expect(geo.defaultPos()).toHaveProperty("coordinates");
  });
  test("check geo Object", () => {
    expect(geo).toHaveProperty("geojson");
    expect(geo).toHaveProperty("shapeLevels");
    expect(geo).toHaveProperty("tile");
    expect(geo).toHaveProperty("getBounds");
    expect(geo).toHaveProperty("defaultPos");
  });
});
