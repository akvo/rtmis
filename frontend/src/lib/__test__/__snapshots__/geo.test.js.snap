// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geo check geo Object 1`] = `
Object {
  "countiesjson": Object {
    "features": Array [],
    "type": "FeatureCollection",
  },
  "defaultPos": [Function],
  "geojson": Object {
    "geometry": null,
    "properties": Object {},
    "type": "Feature",
  },
  "getBounds": [Function],
  "shapeLevels": Array [],
  "tile": Object {
    "attribution": "Tiles &copy; Esri &mdash; DeLorme, NAVTEQ, Esri",
    "url": "https://server.arcgisonline.com/ArcGIS/rest/services/Ocean_Basemap/MapServer/tile/{z}/{y}/{x}",
  },
}
`;
